clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -1094,8 +1094,19 @@ public void handleSuccessfulResponse(RequestHeader requestHeader, long now, Meta
 
             // Check if any topic's metadata failed to get updated
             Map<String, Errors> errors = response.errors();
-            if (!errors.isEmpty())
+            if (!errors.isEmpty()) {
                 log.warn("Error while fetching metadata with correlation id {} : {}", requestHeader.correlationId(), errors);
+                // Print additional debugging information such as leader id and epoch for partitions fail metadata fetch
+                // with UNKNOWN_TOPIC_OR_PARTITION error
+                if (errors.containsKey(Errors.UNKNOWN_TOPIC_OR_PARTITION)) {
+                    List<MetadataResponse.PartitionMetadata> partitionMetadataList = response.topicMetadata().stream()
+                        .flatMap(topicMetadata -> topicMetadata.partitionMetadata().stream()
+                            .filter(partitionMetadata -> partitionMetadata.error() == Errors.UNKNOWN_TOPIC_OR_PARTITION))
+                        .collect(Collectors.toList());
+                    log.warn("Unable to fetch metadata with correlation id {} for topic partitions with these metadata"
```

Comment:
Ideally, we would like to log which broker has returned the response with the errors.
It seems that info is not included in this log.

Reply:
Thanks for the comment Lucas. Instead of adding extra code, I think we can see which broker returned the response by enabling debug log in `NetworkClient#handleCompletedReceives`.
```            
if (log.isDebugEnabled()) {
    log.debug("Received {} response from node {} for request with header {}: {}",
    req.header.apiKey(), req.destination, req.header, response);
}
```
The downside is we might get a lot of other logs in the process. 
It's a bit hard to print the response broker id inside the function (`metadataUpdater.handleSuccessfulResponse()`) where we currently print the log, because response broker id is not passed in as argument or part of `MetadataResponse`.

Reply:
You are right that it's logged there in the debug level logging.
However, the challenge with using that is it's turned off by default for our clients with an INFO level logging.
And when they experience the problem of Unknown topics, the debug level logging above doesn't generate any logs.
By the time they find out the problem, it might have been too late to enable this debug level logging since the problem have been resolved, or their service has been restarted.
Thus we'd like to log the broker id when the error is received.

Can you please dig around to see if it's possible to make that happen by perhaps including the broker id to the MetadataResponse or to a wrapper class including both the broker id and the MetadataResponse? If it's too much work, then we can discuss further.